### PR TITLE
Improve CloudEvents handling and parsing

### LINKED ISSUE #112: Support CloudEvents v0.2
https://github.com/cloudevents/spec/releases/tag/v0.2

### LINKED ISSUE #131: Fix proxy's publish-to-event-stream Kafka test
https://github.com/Accenture/reactive-interaction-gateway/blob/53a41136d9c82cb7d3a07a29c40d675a373db061/apps/rig_tests/test/proxy/publish_to_event_stream/kafka_test.exs#L184

The correlation ID depends on the test's runtime pid, so it's definitely not static.

### PR DESCRIPTION
**Pull request by kevinbader**
This PR's main goal is the implementation of [CloudEvents v0.2](https://github.com/cloudevents/spec/blob/v0.2/spec.md). Additionally, I've reworked how events are represented and handled within RIG in order to allow for a clean implementation, but also to fix some long standing issues.

- The original JSON encoded event is retained to be used for forwarding
the event to subscribers. Internally, the event is processed using an
intermediate representation (parser specific).
- Two different parser implementations (to be benchmarked). For now, the new partial-parser implementation is the default but the standard parser - which pretty much does what the previous code did - is there in case we need it (for performance reasons, or in case we discover a bug in the new parser).
- The CloudEvents implementation turns out to be rather RIG specific;
therefore, the plan of turning it into a standalone library has been
abandoned for now => the new OTP app is `:rig_cloud_events`.

Other stuff:

- Event-*-Tests: I've moved out the SSE and WS clients from the event subscription test to a new module, to be used in other tests.
- The pid serialization is finally fixed - using base64url instead of base64+www-encode.
- With the move to CloudEvents 0.2 I've changed the metadata field names RIG adds in its extension when forwarding a proxy request to a Kafka or Kinesis topic. The new field names are less prone to misspelling/case-issues.

Closes #112.
Closes #131.

### CODE REVIEW

File: apps/rig_cloud_events/test/cloud_events_parser_test.exs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+      # ["bar", "baz"]:
+      assert {:error, {:non_scalar_value, _, _}} = parser.find_value(tokens, ~S"/foo")
+
+      # The list indices syntax is currently not supported/implemented:
+      # assert "bar" = parser.find_value(tokens, ~S"/foo/0")
+
+      assert {:ok, 0} = parser.find_value(tokens, ~S(/))
+      assert {:ok, 1} = parser.find_value(tokens, ~S(/a~1b))
+
+      # Currently a bug in JSONPointer..
```
Comment (kevinbader) [inline apps/rig_cloud_events/test/cloud_events_parser_test.exs]: validate fixed in new version

File: apps/rig_cloud_events/lib/rig_cloud_events/cloud_event.ex [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    else
+      error -> error
+    end
+  end
+
+  @doc """
+  Convenience function used in testing.
+
+  If this would be called in production, of course it would be way more efficient to
+  access the given map directly. However, this modules' raison d'être is the safe
```
[the remark this answers is not in the document]
Comment (kevinbader) [reply apps/rig_cloud_events/lib/rig_cloud_events/cloud_event.ex]: lol

I've seen native speaker writing this a couple of times ¯\_(ツ)_/¯

File: apps/rig/lib/rig/connection/codec.ex [hunk 3]
```diff
@@ -10,8 +10,7 @@ defmodule Rig.Connection.Codec do
   def serialize(pid) do
     pid
     |> :erlang.term_to_binary()
-    |> Base.encode64()
-    |> URI.encode_www_form()
+    |> Base.url_encode64()
 
     # TODO sign this
```
[the remark this answers is not in the document]
Comment (kevinbader) [reply apps/rig/lib/rig/connection/codec.ex]: This is what I've mentioned earlier today - www-form-urlencoding is not path safe but base64url is. So I'd say this is relevant - or do I miss something?

**Comment by kevinbader** [comment]
> Due to removal of eventTime examples now show Invalid time

hm it shouldn't have been _removed_, only _renamed_ to `time`

> smoke failed -- missing COPY apps/rig_cloud_events/mix.exs /opt/sites/rig/apps/rig_cloud_events/ in smoke_tests.dockerfile

*sigh* Okay I'll fix this in another PR and update this PR afterwards..

File: apps/rig_cloud_events/lib/rig_cloud_events/cloud_event.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmacai) [reply apps/rig_cloud_events/lib/rig_cloud_events/cloud_event.ex]: Okay, never saw it :D

**Comment by mmacai** [comment]
> > Due to removal of eventTime examples now show Invalid time
> 
> hm it shouldn't have been _removed_ - only _renamed_ to `time`

but it's removed from RIG or? at least from what I see in changes and is returned to browser

File: apps/rig_cloud_events/mix.exs [hunk 4]
```diff
@@ -24,11 +26,19 @@ defmodule CloudEvents.MixProject do
...
+
   # Run "mix help deps" to learn about dependencies.
   defp deps do
     [
       # Encode/decode to/from JSON:
       {:jason, "~> 1.1"},
+      {:jaxon, "~> 1.0"},
+      # JSON Pointer implementation:
+      {:odgn_json_pointer, "~> 2.3"},
       # Auto-fill eventTime:
```
[the remark this answers is not in the document]
Comment (kevinbader) [reply apps/rig_cloud_events/mix.exs]: re-added event-`time`

[the remark this answers is not in the document]
Comment (mmacai) [reply apps/rig_cloud_events/mix.exs]: based on comment `# Auto-fill eventTime:`

**Comment by mmacai** [comment]
So do we want to do those minor changes to examples separately? If so I can create PR for it.

File: apps/rig_cloud_events/mix.exs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinbader) [reply apps/rig_cloud_events/mix.exs]: aah, now I got it - fixed, nice catch

**Comment by kevinbader** [comment]
> So do we want to do those minor changes to examples separately? If so I can create PR for it.

since they are directly related to the changes in this PR I don't think we need a separate PR... can you please update this one with your changes?

**Pull request merged**